packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,54 +84,58 @@ class OptimisticOracleClient {
     const earliestBlockToQuery = Math.max(currentBlock.number - lookbackBlocks, 0);
 
     // Fetch contract state variables in parallel.
-    const [requestEvents, currentTime] = await Promise.all([
+    const [requestEvents, proposalEvents, disputeEvents, currentTime] = await Promise.all([
       this.oracle.getPastEvents("RequestPrice", { fromBlock: earliestBlockToQuery }),
+      this.oracle.getPastEvents("ProposePrice", { fromBlock: earliestBlockToQuery }),
+      this.oracle.getPastEvents("DisputePrice", { fromBlock: earliestBlockToQuery }),
       this.oracle.methods.getCurrentTime().call()
     ]);
 
-    // Fetch request states in parallel batches, 20 at a time, to be safe and not overload the web3 node.
-    const WEB3_CALLS_BATCH_SIZE = 150;
-
-    const [requestStates] = await Promise.all([
-      Promise.map(
-        requestEvents,
-        request =>
-          // Since we're making an async web3 call here, consider calling
-          // requests(web3.utils.soliditySha3(requester, id, timestamp, acData)) instead
-          // so we can grab more data including that relevant for proposals.
-          this.oracle.methods
-            .getState(
-              request.returnValues.requester,
-              request.returnValues.identifier,
-              request.returnValues.timestamp,
-              request.returnValues.ancillaryData ? request.returnValues.ancillaryData : "0x"
-            )
-            .call(),
-        {
-          concurrency: WEB3_CALLS_BATCH_SIZE
-        }
-      )
-    ]);
+    // Store price requests that have NOT been proposed to yet:
+    const unproposedPriceRequests = requestEvents.filter(event => {
```

Comment:
These array filters/maps are a bit ugly but they are better than making additional web3 calls IMO:

Performance:
- web3 calls: (RequestPrice + ProposePrice + DisputePrice)(`N`)
- Event array computations: (requestEvents + 2 * proposalEvents)(`N`)

Reply:
+1 -- this method of pulling out the state LGTM!

Reply:
So the event Params are slightly different, is there a way to see difference on a subset of the params?